File: java/minimumPathSum.java
class Solution {
    //store the grid as an instance variable
    private int[][] grid;
    //store memo as a map that maps columns onto maps that map row onto subproblem result
    private Map<Integer, Map<Integer, Integer>> memo = new HashMap<>();

    public int minPathSum(int[][] grid) {
        this.grid = grid;
        //add recursive base case to memo
        Map<Integer, Integer> baseCase = new HashMap<>();
        baseCase.put(grid[0].length - 1, grid[grid.length - 1][grid[0].length - 1]);
        memo.put(grid.length - 1, baseCase);
        return recursiveMinPathSum(0, 0);
    }
    
    private int recursiveMinPathSum(int row, int col) {
        //check if problem is in memo
        if (!memo.keySet().contains(row) || !memo.get(row).keySet().contains(col)) {
            //add the current grid to the result
            int result = grid[row][col];
            //add the possible subproblems to a list and add the smallest one to the result
            List<Integer> subProblems = new ArrayList<>();
            if (row < grid.length - 1) {
                subProblems.add(recursiveMinPathSum(row + 1, col));
            }
            if (col < grid[0].length - 1) {
                subProblems.add(recursiveMinPathSum(row, col + 1));
            }
            Collections.sort(subProblems);
            result += subProblems.get(0);
            //add the result to the memo
            if (!memo.keySet().contains(row)) {
                memo.put(row, new HashMap<Integer, Integer>());
            }
            memo.get(row).put(col, result);
        }
        return memo.get(row).get(col);
    }
    
}

File: python/minimum_path_sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        #set the grid to an instance variable
        self._grid = grid
        #add the recursive base class to the memo
        self._memo = { (len(grid) - 1, len(grid[0]) - 1): grid[len(grid) - 1][len(grid[0]) - 1] }
        return self.recursive_min_path_sum(0, 0)
        
    def recursive_min_path_sum(self, row: int, col: int) -> int:
        #check if problem is in memo
        if (row, col) not in self._memo:
            #add the current square to the result
            result = self._grid[row][col]
            #collect the possible sub problems and find the minimum and add it to the result
            sub_problems = []
            if row < len(self._grid) - 1:
                sub_problems.append(self.recursive_min_path_sum(row + 1, col))
            if col < len(self._grid[0]) - 1:
                sub_problems.append(self.recursive_min_path_sum(row, col + 1))
            result += min(sub_problems)
            #add result to memo
            self._memo[(row, col)] = result
        return self._memo[(row, col)]